proposals/20201026-creating-control-plane.md

```diff
@@ -0,0 +1,433 @@
+
+### Goals
+
+- Define the CRD that represents each nested component. The CRD needs to meet two requirements:
+    * Portable - the CRD should hold information that are required by different component providers, e.g., [etcd-operator](https://github.com/coreos/etcd-operator), [etcd-cluster-operator](https://github.com/improbable-eng/etcd-cluster-operator/blob/f84abc6561735814debd67d45bb62d2d2ed8cf4a/api/v1alpha1/etcdcluster_types.go#L31-L47)
+    * Customizable - the CRD should allow end-users to customize each component, i.e., specify the image, component version, and command-line options.
+
+- Define a standard process of creating nested components for NCP.
+
+- Present two examples of creating NCP natively or using the third-party provider to create the NEtcd.
+
+### Non-Goals
+
+- Define how NCP controller works.
+
+- Discuss the details of how the third-party component provider work.
+
+## Proposal
+
+### Portability and Customizability
+
+Generally, creating the three major components requires similar high-level information, like the components' version, the number of replicas, and the amount of computing resources. Meanwhile, end-users should be able to customize NCP components, i.e., specifying the component image, version, and command-line options. Therefore, we define a new struct `NestedComponentSpec` that contains common information required by different providers as well as customized information specified by the end-users. The `NestedComponentSpec` will look like the following
+
+```go
+type NestedComponentSpec struct {
+    // ComponentSpec defines the common information for creating the component
+    // +optional
+    CommonSpec addonv1alpha1.CommonSpec `json:"commonSpec",omitempty`
+    
+    // Resources defines the amount of computing resources that will be used by this component
+    // +optional
+    Resources corev1.ResourceRequirements `json:"resources",omitempty`
+    
+    // Replicas defines the number of replicas in the component's workload 
+    // +optional
+    Replicas int32 `json:"replicas",omitempty`
+
+    // PatchSpecs includes the user specifed settings
+    // +optional
+    PatchSpec addonv1alpha1.PatchSpec `json:"patchSpec,omitempty"`
```

Comment:
I'd just inline this as well, instead of having a different struct

Reply:
Oh thanks, my updates were wrong. 🤦‍♂️ (edited)

Reply:
Oh, I actually meant to drop `PatchSpec` altogether if possible and just have the fields in here, it'd be easier to use

Reply:
We can, although I'd like to think about actually supporting all the patching/mutation funcs from https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/ so that we don't have to immediately expose all the different controls for mutating the individual components.

For example my apiservers might need to turn on `DualStack` or another featuregate which might not be exposed as a typed field in the spec just yet, this allows us to be more flexible upfront and overtime "graduate" fields to the top-level component CRs. 

I have an item on the ClusterAddons agenda next week (https://docs.google.com/document/d/10_tl_SXcFGb-2109QpcFVrdrfnVEuQ05MBrXtasB0vk/edit#heading=h.8qktiog8x5ri) to bring up our use cases and gather any concerns they might have, do you think allowing this upfront concerning?